src/foundations/type/Type.tsx

```diff
@@ -0,0 +1,124 @@
+import styled, { css } from 'styled-components';
+import { FlattenSimpleInterpolation } from 'styled-components/ts3.6';
+
+import { Utils } from '../../utils';
+
+import { BODY_SIZES, DEFAULT_TYPE_SIZE, HEADING_SIZES, HEADING_ELEMENTS } from './Type.constants';
+import { PType } from './Type.props';
+import { TTypeSize, TTypeType } from './Type.types';
+
+function generateFontStyleVariations(
+    typeSize: TTypeSize,
+    fontType: TTypeType
+): FlattenSimpleInterpolation {
+    return css`
+        &[data-size='${typeSize}'] {
+            font-size: var(--${fontType}-font-size-${typeSize});
+            line-height: var(--${fontType}-line-height-${typeSize});
+
+            &[data-margin='none'] {
+                margin: 0;
+            }
+            &[data-margin='top'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0 0;
+            }
+            &[data-margin='bottom'] {
+                margin: 0 0 var(--${fontType}-margin-bottom-${typeSize});
+            }
+            &[data-margin='both'] {
+                margin: var(--${fontType}-margin-top-${typeSize}) 0
+                    var(--${fontType}-margin-bottom-${typeSize});
+            }
+        }
+    `;
+}
+
+const Type = styled.p
+    .attrs((props: PType) => ({
+        // it is possible to remap props, so we do not need to pass down the
+        // `as` property from styled-components and prevent usage of
+        // unsupported HTML tags
+        as: props.element,
+        'data-size': props.size || DEFAULT_TYPE_SIZE,
+        'data-type': HEADING_ELEMENTS.includes(props.element) ? 'heading' : 'body',
+        'data-weight': props.weight,
+        'data-margin': props.margin,
+        'data-color': props.color || 'primary',
```

Comment:
Should we also use defaults constants for all of these like with `data-size` and the `Heading` and `Text` components?

Reply:
no, we should replace those entirely, since we already set default values on the components